File: cpp/game_panel.cpp
// SDL2 rendering implementation for Pacman.
// Source of truth: shared/game_constants.json
#include "game_panel.h"
#include "game_logic.h"
#include <cstdlib>
#include <string>
#include <sstream>

// ---------------------------------------------------------------------------
// Construction / destruction
// ---------------------------------------------------------------------------

GamePanel::GamePanel(GameLogic& logic, SDL_Renderer* renderer)
    : gameLogic(logic),
      renderer(renderer),
      running(true)
{
}

GamePanel::~GamePanel() {
    // We do NOT own the renderer — caller destroys it.
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

bool GamePanel::isRunning() const {
    return running;
}

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Input Handling — SDL2 Event Processing
// ═══════════════════════════════════════════════════════
// SDL2 uses a C-style event union (SDL_Event) where the `type` field tells you
// which member of the union is valid. For SDL_KEYDOWN, `event.key.keysym.scancode`
// gives the physical key pressed. The switch/case maps scancodes to game
// directions. SDL_SCANCODE_* refers to physical key positions (layout-independent),
// while SDL_KEYCODE_* (SDLK_*) refers to logical characters — scancodes are
// preferred for game input since WASD stays in the same place on any keyboard.
//
// Compare with:
//   Java → `KeyListener.keyPressed(KeyEvent e)` with `e.getKeyCode()`.
//   Python → Pygame: `pygame.event.get()` with `event.key == pygame.K_UP`.
//   JavaScript → `document.addEventListener('keydown', e => e.key)`.
//   TypeScript → Same as JS with `KeyboardEvent` type for the event object.
//
// KEY CONCEPT: SDL2 scancodes represent physical key positions — use them
// for game input so controls work correctly on any keyboard layout.
// ═══════════════════════════════════════════════════════
void GamePanel::handleInput(const SDL_Event& event) {
    if (event.type == SDL_QUIT) {
        running = false;
        return;
    }

    if (event.type == SDL_KEYDOWN) {
        switch (event.key.keysym.scancode) {
            case SDL_SCANCODE_UP:
                gameLogic.getPacman().setDirection(0, -1);
                break;
            case SDL_SCANCODE_DOWN:
                gameLogic.getPacman().setDirection(0, 1);
                break;
            case SDL_SCANCODE_LEFT:
                gameLogic.getPacman().setDirection(-1, 0);
                break;
            case SDL_SCANCODE_RIGHT:
                gameLogic.getPacman().setDirection(1, 0);
                break;
            case SDL_SCANCODE_ESCAPE:
                running = false;
                break;
            default:
                break;
        }
    }
}

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Rendering — SDL2 Draw-Then-Present Pipeline
// ═══════════════════════════════════════════════════════
// SDL2 uses double-buffering: all draw calls (SetRenderDrawColor, FillRect,
// etc.) write to a back buffer. SDL_RenderPresent() then swaps it to the
// screen in one atomic operation, preventing flicker. This is the same
// principle behind Java's BufferStrategy, Pygame's display.flip(), and the
// HTML5 Canvas (which auto-presents after the JS event handler returns).
//
// Compare with:
//   Java → `BufferStrategy.show()` or Swing's automatic double-buffering.
//   Python → Pygame `display.flip()` or `display.update()`.
//   JavaScript → Canvas auto-presents; WebGL needs explicit `gl.flush()`.
//   TypeScript → Same as JS; canvas rendering is identical.
//
// KEY CONCEPT: Double-buffering draws to an offscreen buffer then flips it
// to the display — preventing visual tearing and partial-frame artifacts.
// ═══════════════════════════════════════════════════════
void GamePanel::render() {
    // Clear to black background
    SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
    SDL_RenderClear(renderer);

    renderMaze();
    renderPellets();
    renderPacman();
    renderGhosts();

    if (gameLogic.isGameOver()) {
        renderGameOver();
    }

    SDL_RenderPresent(renderer);
}

void GamePanel::updateTitle(SDL_Window* window) const {
    std::string title = "Pacman - C++/SDL2  |  Score: "
                      + std::to_string(gameLogic.getScore());
    if (gameLogic.isGameOver()) {
        title += "  |  GAME OVER";
    }
    SDL_SetWindowTitle(window, title.c_str());
}

// ---------------------------------------------------------------------------
// Rendering helpers
// ---------------------------------------------------------------------------

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Rendering — Tile-Based Drawing with SDL_Rect
// ═══════════════════════════════════════════════════════
// Each maze cell is drawn as a TILE_SIZE x TILE_SIZE filled rectangle.
// SDL_Rect is a plain C struct {int x, y, w, h} — no constructor, no
// destructor, zero overhead. C++ allows aggregate initialization with braces:
// `SDL_Rect rect = {x, y, w, h};` which is concise and efficient. The nested
// loop iterates all cells, drawing only walls — this brute-force approach
// works fine for small grids but would need culling/spatial partitioning for
// large worlds.
//
// Compare with:
//   Java → `Graphics2D.fillRect(x, y, w, h)` — similar immediate-mode API.
//   Python → `pygame.draw.rect(surface, color, (x, y, w, h))`.
//   JavaScript → `ctx.fillRect(x, y, w, h)` — Canvas 2D is very similar.
//   TypeScript → Same as JS Canvas API with typed context.
//
// KEY CONCEPT: SDL_Rect is a plain-old-data (POD) struct with zero overhead —
// C++ aggregate initialization makes creating them concise and allocation-free.
// ═══════════════════════════════════════════════════════
void GamePanel::renderMaze() {
    const Maze& maze = gameLogic.getMaze();

    // Walls: blue filled rectangles, TILE_SIZE x TILE_SIZE
    SDL_SetRenderDrawColor(renderer, 0, 0, 255, 255);

    for (int x = 0; x < maze.getWidth(); ++x) {
        for (int y = 0; y < maze.getHeight(); ++y) {
            if (maze.isWall(x, y)) {
                SDL_Rect rect = {
                    x * TILE_SIZE,
                    y * TILE_SIZE,
                    TILE_SIZE,
                    TILE_SIZE
                };
                SDL_RenderFillRect(renderer, &rect);
            }
        }
    }
}

void GamePanel::renderPellets() {
    const Maze& maze = gameLogic.getMaze();

    // Pellets: white 4x4 rectangles centered in each tile
    SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);

    const int pelletSize = 4;
    const int offset = (TILE_SIZE - pelletSize) / 2;

    for (int x = 0; x < maze.getWidth(); ++x) {
        for (int y = 0; y < maze.getHeight(); ++y) {
            if (maze.hasPellet(x, y)) {
                SDL_Rect rect = {
                    x * TILE_SIZE + offset,
                    y * TILE_SIZE + offset,
                    pelletSize,
                    pelletSize
                };
                SDL_RenderFillRect(renderer, &rect);
            }
        }
    }
}

void GamePanel::renderPacman() {
    const Pacman& pac = gameLogic.getPacman();

    // Yellow 16x16 filled rectangle centered in the tile
    SDL_SetRenderDrawColor(renderer, 255, 255, 0, 255);

    const int pacSize = 16;
    const int offset = (TILE_SIZE - pacSize) / 2;

    SDL_Rect rect = {
        pac.getGridX() * TILE_SIZE + offset,
        pac.getGridY() * TILE_SIZE + offset,
        pacSize,
        pacSize
    };
    SDL_RenderFillRect(renderer, &rect);
}

void GamePanel::renderGhosts() {
    const auto& ghosts = gameLogic.getGhosts();

    const int ghostSize = 18;
    const int bodyOffset = (TILE_SIZE - ghostSize) / 2;
    const int eyeSize = 4;

    for (const auto& ghost : ghosts) {
        Uint8 r, g, b;
        parseHexColor(ghost.getColor(), r, g, b);

        // Ghost body: colored 18x18 rectangle
        SDL_SetRenderDrawColor(renderer, r, g, b, 255);
        SDL_Rect body = {
            ghost.getGridX() * TILE_SIZE + bodyOffset,
            ghost.getGridY() * TILE_SIZE + bodyOffset,
            ghostSize,
            ghostSize
        };
        SDL_RenderFillRect(renderer, &body);

        // Eyes: two white 4x4 rectangles in upper portion of the ghost body
        SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);

        // Left eye
        SDL_Rect leftEye = {
            body.x + 3,
            body.y + 3,
            eyeSize,
            eyeSize
        };
        SDL_RenderFillRect(renderer, &leftEye);

        // Right eye
        SDL_Rect rightEye = {
            body.x + ghostSize - 3 - eyeSize,
            body.y + 3,
            eyeSize,
            eyeSize
        };
        SDL_RenderFillRect(renderer, &rightEye);

        // Pupils: two dark-blue 2x2 rectangles inside the eyes
        SDL_SetRenderDrawColor(renderer, 0, 0, 128, 255);

        SDL_Rect leftPupil = {
            leftEye.x + 1,
            leftEye.y + 1,
            2,
            2
        };
        SDL_RenderFillRect(renderer, &leftPupil);

        SDL_Rect rightPupil = {
            rightEye.x + 1,
            rightEye.y + 1,
            2,
            2
        };
        SDL_RenderFillRect(renderer, &rightPupil);
    }
}

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Rendering — Alpha Blending for Overlays
// ═══════════════════════════════════════════════════════
// SDL2 requires explicitly enabling blend mode before drawing translucent
// shapes. The alpha channel (120 out of 255 here) controls opacity. After
// drawing, blend mode is reset to NONE for subsequent opaque draws. This
// explicit state management is typical of C-level graphics APIs where you
// control the GPU pipeline state directly.
//
// Compare with:
//   Java → `Graphics2D.setComposite(AlphaComposite.getInstance(...))`.
//   Python → Pygame: `surface.set_alpha(120)` or per-pixel alpha with convert_alpha().
//   JavaScript → `ctx.globalAlpha = 0.47` or `rgba(200, 0, 0, 0.47)` in fillStyle.
//   TypeScript → Same as JS Canvas; `globalAlpha` property.
//
// KEY CONCEPT: SDL2's explicit blend-mode state must be toggled on/off —
// forgetting to reset it is a common bug that makes everything translucent.
// ═══════════════════════════════════════════════════════
void GamePanel::renderGameOver() {
    // Draw a semi-transparent red overlay to indicate game over.
    // SDL2 supports alpha blending with SDL_BLENDMODE_BLEND.
    SDL_SetRenderDrawBlendMode(renderer, SDL_BLENDMODE_BLEND);
    SDL_SetRenderDrawColor(renderer, 200, 0, 0, 120);

    SDL_Rect overlay = { 0, 0, WINDOW_WIDTH, WINDOW_HEIGHT };
    SDL_RenderFillRect(renderer, &overlay);

    // Reset blend mode
    SDL_SetRenderDrawBlendMode(renderer, SDL_BLENDMODE_NONE);

    // Draw a centered white rectangle as a "GAME OVER" placeholder banner.
    // (Proper text would require SDL_ttf; the title bar already shows the text.)
    SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);
    const int bannerW = 200;
    const int bannerH = 40;
    SDL_Rect banner = {
        (WINDOW_WIDTH - bannerW) / 2,
        (WINDOW_HEIGHT - bannerH) / 2,
        bannerW,
        bannerH
    };
    SDL_RenderFillRect(renderer, &banner);

    // Inner dark rectangle to create a border effect
    SDL_SetRenderDrawColor(renderer, 200, 0, 0, 255);
    SDL_Rect inner = {
        banner.x + 4,
        banner.y + 4,
        bannerW - 8,
        bannerH - 8
    };
    SDL_RenderFillRect(renderer, &inner);
}

// ---------------------------------------------------------------------------
// Utility
// ---------------------------------------------------------------------------

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: String Parsing — istringstream for Hex Colors
// ═══════════════════════════════════════════════════════
// C++ has no built-in hex-color parser, so we use std::istringstream with
// the `std::hex` manipulator to parse "#RRGGBB" into an unsigned int, then
// extract R/G/B via bit-shifting. The static_cast<Uint8> is required because
// C++ is strict about narrowing conversions. This manual parsing is verbose
// compared to other languages but gives full control over error handling.
//
// Compare with:
//   Java → `Color.decode("#FF0000")` or `Integer.parseInt(hex, 16)`.
//   Python → `int(hex[1:], 16)` — one-liner with slice and base conversion.
//   JavaScript → `parseInt(hex.slice(1), 16)` — very similar to Python.
//   TypeScript → Same as JS; could add a return type like `{r: number, g: number, b: number}`.
//
// KEY CONCEPT: C++ requires explicit parsing and casting for type conversions —
// more verbose than dynamic languages but catches type errors at compile time.
// ═══════════════════════════════════════════════════════
void GamePanel::parseHexColor(const std::string& hex, Uint8& r, Uint8& g, Uint8& b) {
    // Expected format: "#RRGGBB"
    if (hex.size() == 7 && hex[0] == '#') {
        unsigned int rgb = 0;
        std::istringstream iss(hex.substr(1));
        iss >> std::hex >> rgb;
        r = static_cast<Uint8>((rgb >> 16) & 0xFF);
        g = static_cast<Uint8>((rgb >> 8) & 0xFF);
        b = static_cast<Uint8>(rgb & 0xFF);
    } else {
        // Fallback: white
        r = 255;
        g = 255;
        b = 255;
    }
}

File: python/game_panel.py
# Source of truth: shared/game_constants.json

import pygame

from pacman import Pacman
from ghost import Ghost
from game_logic import GameLogic

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------
PANEL_WIDTH = 800
PANEL_HEIGHT = 600
TILE_SIZE = 20
FPS = 60

# ═══════════════════════════════════════════════════════
# EDUCATIONAL NOTE: Rendering/Drawing
# ═══════════════════════════════════════════════════════
# Colors are plain tuples of (R, G, B) integers. Python tuples
# are immutable and lightweight -- perfect for constants that
# never change. No need for a Color class or static final fields.
#
# Compare with:
#   Java (Swing) → Color.BLUE, Color.WHITE, new Color(255,255,0)
#   Java uses Color objects from java.awt. Python tuples are
#   simpler and do not require any import -- they are built-in.
#
# KEY CONCEPT: Python tuples serve as lightweight, immutable
# data containers -- ideal for constants like RGB colors.
# ═══════════════════════════════════════════════════════

# Colors
BLACK = (0, 0, 0)
BLUE = (0, 0, 255)
WHITE = (255, 255, 255)
YELLOW = (255, 255, 0)


class GamePanel:
    """Rendering and input layer for Pacman.

    Owns the pygame display surface and clock. All rendering is centralised
    here -- Pacman, Ghost, Maze, and GameLogic are pure data/logic classes
    with ZERO pygame imports.
    """

    def __init__(self):
        self._screen: pygame.Surface = pygame.display.set_mode(
            (PANEL_WIDTH, PANEL_HEIGHT)
        )
        pygame.display.set_caption("Pacman - Python/Pygame")
        self._clock: pygame.time.Clock = pygame.time.Clock()
        self._font: pygame.font.Font = pygame.font.SysFont(None, 28)
        self._game_over_font: pygame.font.Font = pygame.font.SysFont(None, 64)
        self._running: bool = True

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def running(self) -> bool:
        return self._running

    @property
    def screen(self) -> pygame.Surface:
        return self._screen

    @property
    def clock(self) -> pygame.time.Clock:
        return self._clock

    # ------------------------------------------------------------------
    # Input handling
    # ------------------------------------------------------------------

    def handle_input(self, pacman: Pacman) -> None:
        """Process pygame events.

        Arrow keys set Pacman's buffered direction.
        Closing the window (or pressing Escape) stops the game loop.
        """
        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Input Handling
        # ═══════════════════════════════════════════════════════
        # Pygame uses an EVENT QUEUE: all keyboard/mouse/window events
        # are collected by the OS and retrieved with pygame.event.get().
        # We iterate through ALL events each frame and respond to the
        # ones we care about (QUIT, KEYDOWN). Unhandled events are
        # silently ignored -- no crash, no error.
        #
        # Compare with:
        #   Java (Swing) → addKeyListener(new KeyAdapter() { ... })
        #   Java registers a listener OBJECT that the framework calls
        #   back asynchronously. Python/pygame polls events explicitly
        #   in the game loop -- simpler flow, no callback surprises.
        #
        # KEY CONCEPT: Polling (check every frame) vs. Callbacks
        # (framework calls you). Pygame uses polling, Swing uses
        # callbacks. Polling gives you full control over timing.
        # ═══════════════════════════════════════════════════════

        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                self._running = False
            elif event.type == pygame.KEYDOWN:
                if event.key == pygame.K_LEFT:
                    pacman.set_direction(-1, 0)
                elif event.key == pygame.K_RIGHT:
                    pacman.set_direction(1, 0)
                elif event.key == pygame.K_UP:
                    pacman.set_direction(0, -1)
                elif event.key == pygame.K_DOWN:
                    pacman.set_direction(0, 1)
                elif event.key == pygame.K_ESCAPE:
                    self._running = False

    # ------------------------------------------------------------------
    # Update
    # ------------------------------------------------------------------

    def update(
        self,
        pacman: Pacman,
        ghosts: list[Ghost],
        game_logic: GameLogic,
    ) -> None:
        """Advance game state by one frame (skip if game over)."""
        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Object Movement
        # ═══════════════════════════════════════════════════════
        # The update order matters: Pacman moves first, then each
        # ghost moves, then collisions are checked. This ensures
        # that collision detection always sees the FINAL positions
        # for this frame. Changing the order (e.g., checking
        # collisions before ghosts move) would create subtle bugs.
        #
        # Compare with:
        #   Java (Swing) → actionPerformed(ActionEvent e) in a
        #   Timer callback follows the same sequence. Python makes
        #   the ordering more visible because the calls are plain
        #   sequential statements -- no event-dispatch indirection.
        #
        # KEY CONCEPT: Update order (move -> check collisions) is
        # a critical game-loop design decision. Always be explicit.
        # ═══════════════════════════════════════════════════════

        if game_logic.game_over:
            return

        pacman.update(game_logic.maze)

        for ghost in ghosts:
            ghost.update(game_logic.maze)

        game_logic.check_collisions(pacman, ghosts)

    # ------------------------------------------------------------------
    # Drawing
    # ------------------------------------------------------------------

    def draw(
        self,
        screen: pygame.Surface,
        pacman: Pacman,
        ghosts: list[Ghost],
        game_logic: GameLogic,
    ) -> None:
        """Render the entire frame: maze, entities, HUD."""
        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Rendering/Drawing
        # ═══════════════════════════════════════════════════════
        # Every frame starts by filling the screen with BLACK (clearing
        # the previous frame), then redraws everything from scratch:
        # walls, pellets, Pacman, ghosts, HUD, and game-over overlay.
        # This is called "immediate mode" rendering -- nothing persists
        # between frames; you redraw the entire scene each time.
        #
        # Compare with:
        #   Java (Swing) → paintComponent(Graphics g) works the same
        #   way: clear with g.setColor(BLACK); g.fillRect(...), then
        #   redraw. Java uses Graphics2D methods; Python uses pygame's
        #   draw module (pygame.draw.rect, pygame.draw.circle).
        #
        # KEY CONCEPT: Immediate-mode rendering (clear + redraw every
        # frame) is standard for 2D games in both Python and Java.
        # ═══════════════════════════════════════════════════════

        screen.fill(BLACK)

        maze = game_logic.maze
        maze_w = maze.width
        maze_h = maze.height

        # --- Walls (blue filled rectangles) ---
        for x in range(maze_w):
            for y in range(maze_h):
                if maze.is_wall(x, y):
                    pygame.draw.rect(
                        screen,
                        BLUE,
                        (x * TILE_SIZE, y * TILE_SIZE, TILE_SIZE, TILE_SIZE),
                    )

        # --- Pellets (white circles, 4px diameter, centred in tile) ---
        for x in range(maze_w):
            for y in range(maze_h):
                if maze.has_pellet(x, y):
                    center_x = x * TILE_SIZE + TILE_SIZE // 2
                    center_y = y * TILE_SIZE + TILE_SIZE // 2
                    pygame.draw.circle(screen, WHITE, (center_x, center_y), 2)

        # --- Pacman (yellow filled circle, full tile size) ---
        pac_center_x = pacman.grid_x * TILE_SIZE + TILE_SIZE // 2
        pac_center_y = pacman.grid_y * TILE_SIZE + TILE_SIZE // 2
        pygame.draw.circle(
            screen, YELLOW, (pac_center_x, pac_center_y), TILE_SIZE // 2
        )

        # --- Ghosts (coloured rectangle + two white eye circles) ---
        for ghost in ghosts:
            color = (ghost.color_r, ghost.color_g, ghost.color_b)
            gx = ghost.grid_x * TILE_SIZE
            gy = ghost.grid_y * TILE_SIZE

            # Body
            pygame.draw.rect(
                screen, color, (gx, gy, ghost.size, ghost.size)
            )

            # Eyes (two white circles matching the Java reference offsets)
            pygame.draw.circle(screen, WHITE, (gx + 7, gy + 7), 2)
            pygame.draw.circle(screen, WHITE, (gx + 13, gy + 7), 2)

        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Rendering/Drawing
        # ═══════════════════════════════════════════════════════
        # Text in pygame is a two-step process: (1) render the string
        # to a Surface with font.render(), (2) blit (copy) that Surface
        # onto the screen. pygame.display.flip() then pushes the
        # completed frame to the monitor in one operation (double
        # buffering prevents screen tearing).
        #
        # Compare with:
        #   Java (Swing) → g.drawString("Score: " + score, x, y)
        #   Java draws text directly onto the Graphics context.
        #   Pygame's Surface-based approach is more explicit but
        #   gives you precise control over text positioning via
        #   get_rect() and centering helpers.
        #
        # KEY CONCEPT: Double buffering (draw offscreen, then flip)
        # eliminates flicker and tearing in both pygame and Swing.
        # ═══════════════════════════════════════════════════════

        # --- Score HUD (white text, bottom-left) ---
        score_surface = self._font.render(
            f"Score: {game_logic.score}", True, WHITE
        )
        screen.blit(score_surface, (10, PANEL_HEIGHT - 30))

        # --- Game Over overlay ---
        if game_logic.game_over:
            go_surface = self._game_over_font.render("GAME OVER", True, WHITE)
            go_rect = go_surface.get_rect(
                center=(PANEL_WIDTH // 2, PANEL_HEIGHT // 2)
            )
            screen.blit(go_surface, go_rect)

        pygame.display.flip()
